core/hooks/Model.js

```diff
@@ -0,0 +1,114 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2019 Extremely Heavy Industries Inc.
+ */
+
+import {useState, useContext} from 'react';
+import {useOnMount, useOnUnmount} from './Mount';
+import {isPlainObject} from 'lodash';
+import {XH} from '@xh/hoist/core/index';
+import {RefreshContext} from '@xh/hoist/core/refresh/RefreshContext';
+
+
+/**
+ * Hook supporting components that will receive a model from props. The model can be passed as
+ * either an already-created class instance or as a config for one to be created internally.
+ *
+ * Models that are created internally by the component may be considered "owned" models. This hook
+ * will ensure an owned model's destroy() method is called when its component is unmounted.
+ *
+ * Parent components should provide concrete instances of models to their children only if they wish
+ * to programmatically access those models to reference data or otherwise manipulate the component
+ * using the model's API. Otherwise, models should be specified via a config object to take
+ * advantage of the auto-cleanup process.
+ *
+ * The model instance is not expected to change for the lifetime of the component. Apps that wish
+ * to swap out the model for a mounted component should ensure that a new instance of the component
+ * gets mounted. This can be done by setting the component's `key` prop to `model.xhId`, as
+ * HoistModels always return IDs unique to each instance.
+ *
+ * @param {Object} modelClass - class defining a HoistModel
+ * @param {Array} props
+ * @returns {Object} - instance of a HoistModel
+ */
+export function useProvidedModel(modelClass, props) {
+    let [state] = useState({instance: null, isOwned: false}),
+        propsModel = props.model;
+
+    // Recognize any disallowed dynamic change to the model
+    if (state.instance && !state.isOwned && propsModel !== state.instance) throwModelChangeException();
+
+    // If we don't yet have a model, potentially source from props, instantiating if appropriate.
+    if (!state.instance && propsModel) {
+        if (isPlainObject(propsModel)) {
+            if (modelClass) {
+                state.instance = new modelClass(propsModel);
+                state.isOwned = true;
```

Comment:
As discussed on IM today, wary of mutating state directly here. Should we revisit with `useRef()`?